packages/perseus/src/util/snowman-utils.ts

```diff
@@ -0,0 +1,157 @@
+export enum WidgetType {
+    Categorizer = "categorizer",
+    CSProgram = "cs-program",
+    Definition = "definition",
+    Dropdown = "dropdown",
+    Explanation = "explanation",
+    Expression = "expression",
+    GradedGroup = "graded-group",
+    GradedGroupAnswerBar = "graded-group-answer-bar",
+    GradedGroupSet = "graded-group-set",
+    Grapher = "grapher",
+    Group = "group",
+    Iframe = "iframe",
+    Image = "image",
+    InputNumber = "input-number",
+    Interaction = "interaction",
+    InteractiveGraph = "interactive-graph",
+    InteractiveNumberLine = "interactive-number-line",
+    LabelImage = "label-image",
+    Matcher = "matcher",
+    Matrix = "matrix",
+    Measurer = "measurer",
+    Molecule = "molecule",
+    NumberLine = "number-line",
+    NumericInput = "numeric-input",
+    Orderer = "orderer",
+    Passage = "passage",
+    PassageRef = "passage-ref",
+    PassageRefTarget = "passage-ref-target",
+    Plotter = "plotter",
+    Protractor = "protractor",
+    PythonProgram = "python-program",
+    Radio = "radio",
+    Sorter = "sorter",
+    Table = "table",
+    Unit = "unit",
+    Video = "video",
+    VideoTranscriptLink = "video-transcript-link",
+}
+
+export type WidgetId = `${WidgetType} ${number}`;
+
+// These are all the current widgets that require user-input and are supported for automatic scoring
+export const QUESTION_WIDGETS: ReadonlyArray<WidgetType> = [
+    WidgetType.NumericInput,
+    WidgetType.InputNumber,
+    WidgetType.Radio,
+    WidgetType.InteractiveGraph,
+    WidgetType.InteractiveNumberLine,
+    WidgetType.Categorizer,
+    WidgetType.Plotter,
+    WidgetType.Orderer,
+    WidgetType.Protractor,
+    WidgetType.Matcher,
+    WidgetType.Sorter,
+];
+
+/**
+ * Regex for widget placeholders in a string.
+ *
+ * First capture group is the widget ID (ex. 'radio 1')
+ * Second capture group is the widget type (ex. "radio)
+ * exec return will look like: ['[[☃ radio 1]]', 'radio 1', 'radio']
+ */
+export const widgetRegex = /\[\[☃ (([a-z-]+) \d+)\]\]/g;
```

Comment:
I think this might be a bug. Since this is declared globally and adding the `g` makes it stateful, I think it's going to cause erratic behavior:

``` JS
const regexWithG = /\[(\w+) (\d+)\]/g
console.log(regexWithG.exec("[hello 42] [world 5]"))
// ["[hello 42]", "hello", "42"]

console.log(regexWithG.exec("[something 42] [else 5]"))
// ["[else 5]", "else", "5"]

console.log(regexWithG.exec("[new 42] [thing 5]"))
// null
```

Reply:
For sure. The webapp instances where I took the regex from all use a while loop to specifically loop through all matches of the regex, so they want it to be stateful. Not sure how to fix it so that the lastIndex property resets after each use other than to have each user make sure they reset it. Is there something else we could do here? We could make it not global and then refactor wherever it uses a for loop to somehow take that into account?

Reply:
Going to use the generator function for this! Thanks!